README.md
# hs-notifications

A Super Simple Desktop Notification Server in Haskell.

![A Screenshot Showing the Notification Windows Created by hs-notifications](http://bugs.sleepanarchy.com/projects/hs-notifications/repository/revisions/master/entry/screenshot.png  "hs-notifications Screenshot")


`DBUS_SESSION_BUS_ADDRESS` needs to be set, which can be accomplished by adding
`export $(dbus-launch)` to your `~/.xinitrc`.


Install `stack` & `gobject-introspection`:

    sudo pacman -S stack gobject-introspection

Build:

    stack build

Run:

    stack exec hs-notifications

Install to `~/.local/bin`:

    stack install
    export PATH="$HOME/.local/bin/:$PATH"
    hs-notifications

`Mod4-Shift-Ctrl-w` removes all the notifications. `Mod4-Ctrl-w` removes the
oldest notification. You can also click them.

You can change the colors & keybindings by creating a
`~/.config/hs-notifications/config.ini`, see the `config-example.ini` for
available options.

My plan is to import this into my xmond config, build my notification Config in
there, & use `forkProcess` to run the application.

I've built this to my needs - if you would like to use this but want some more
features or customizability, feel free to open an issue.


## Architecture

This was hacked together in a day or so, there's lots of room for
improvement.

There's a DBus client & a GTK app. Main just initializes the shared state,
forks the DBus client and starts the GTk app.

### Init

The DBus client connects to the current dbus session, identifies itself as the
notification server, and sets up it's event handlers.

The GTK app initializes itself, gets the notification placement position,
attaches some CSS sytles, and sets up some keybindings & timers.

### New Notification

When the DBus client receives a notification, it adds it to the notification
queue in the shared state.

Every second, the GTK loop checks the notification queue for any new messages.
Any messages are rendered & added to the window list.

As windows are added, the application tracks the position of the next window by
adding the new window's height & some spacing.

### Removing Notifications

You can click notifications to remove them, press `Mod4-Ctrl-w` to remove
the oldest, or press `Mod4-Ctrl-Shift-w` to remove them all.

Removing a notification updates the window list, re-adjusts the positions of
windows below it, & adds its `NotificationID` to the removal queue.

The DBus client wakes up every second to check the removal queue & send the
`NotificationClosed` signal for any removed notifications.


## License

GPL-3.0


ChangeLog.md
# Changelog for hs-notifications

## Unreleased changes


src/Lib.hs
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE OverloadedStrings #-}
module Lib where

import Control.Applicative ((<|>))
import Control.Arrow (first, second)
import Control.Concurrent (threadDelay, forkIO, killThread)
import Control.Concurrent.STM (TVar, atomically, newTVarIO, modifyTVar, readTVar, readTVarIO, writeTVar)
import Control.Exception (bracket)
import Control.Monad ((<=<), (>=>), forever, when, void, join, foldM)
import Data.Bits ((.|.))
import Data.Int (Int32)
import Data.Time.Clock (UTCTime, getCurrentTime, addUTCTime)
import Data.List (partition, subsequences)
import Data.Maybe (listToMaybe)
import Data.Monoid ((<>))
import Data.Word (Word32)
import System.Exit (exitFailure)

import DBus
import DBus.Client

import qualified Data.Map as M
import qualified Data.Text as T
import qualified GI.Gtk as Gtk
import qualified GI.Gdk as Gdk
import qualified GI.GLib as GLib
import qualified Graphics.X11.Xlib as Xlib


notificationSpacing :: Int32
notificationSpacing = 5

offsetX :: Int32
offsetX = 12
offsetY :: Int32
offsetY = 20
verticalPadding :: Int32
verticalPadding = 10
horizontalPadding :: Int32
horizontalPadding = 20

-- Combine GTK + DBUS
run :: IO ()
run =
    initialState >>= \s -> forkIO (connectAndServe s) >> runGtk s


data AppState
    = AppState
        { appNotificationQueue :: [Notification]    -- ^ Newest notifications at front
        , appRemovalQueue :: [(NotificationID, ReasonClosed)]
        , appWindowList :: [(Notification, Gtk.Window)] -- ^ Add new windows to end
        , appRootPosition :: (Int32, Int32)
        , appNextPosition :: (Int32, Int32)
        , appNextNotificationID :: NotificationID
        }

initialState :: IO (TVar AppState)
initialState =
    newTVarIO AppState
        { appNotificationQueue =
            []
        , appRemovalQueue =
            []
        , appWindowList =
            []
        , appRootPosition =
            ( 0, 0 )
        , appNextPosition =
            ( 0, 0 )
        , appNextNotificationID =
            NotificationID minBound
        }


newtype NotificationID
    = NotificationID
        { fromNotificationID :: Word32
        } deriving (Eq)

nextID :: NotificationID -> NotificationID
nextID (NotificationID w) =
    if w == maxBound then
        NotificationID minBound
    else
        NotificationID $ succ w


data Notification
    = Notification
        { nTitle :: T.Text
        , nBody :: T.Text
        , nID :: NotificationID
        , nExpirationTime :: Maybe UTCTime
        }


-- TODO: Pull all the constants & colors into a Config type
-- TODO: Make left & right notification lists
-- TODO: Make customizable format strings matching on app name
-- TODO: Refactor so this is just like 5 function calls
runGtk :: TVar AppState -> IO ()
runGtk sTV = do
    -- Initialization
    void $ Gtk.init Nothing

    rootPosition <- first (+ offsetX) . second (+ offsetY)
        <$> getMonitorGeometryOrExit
    atomically . modifyTVar sTV $ \s -> s
        { appRootPosition = rootPosition
        , appNextPosition = rootPosition
        }


    -- Attach Border Color Style
    -- TODO: Separate function - do more of the themeing in css, generate css w/ library?
    Gdk.screenGetDefault >>= \case
        Nothing ->
            return ()
        Just s -> do
            provider <- Gtk.cssProviderNew
            Gtk.cssProviderLoadFromData provider
                "window { border-color: #F92672; border-style: solid; border-width: 1px; }"
            Gtk.styleContextAddProviderForScreen s provider 401


    -- Keybind Watchers

    -- TODO: Bind Global Clear All Key
    -- TODO: Move to separate function
    -- All these Xlib calls can throw exceptions...
    shortcutThreadId <- forkIO . bracket (Xlib.openDisplay "") Xlib.closeDisplay $ \display -> do
        let rootWindow = Xlib.defaultRootWindow display
        wKeyCode <- Xlib.keysymToKeycode display Xlib.xK_w
        Xlib.selectInput display Xlib.keyPressMask rootWindow
        void . Xlib.allocaXEvent $ \ev ->
            grabKeyWithIgnoreMasks display wKeyCode (Xlib.mod4Mask .|. Xlib.controlMask)
                rootWindow False Xlib.grabModeAsync Xlib.grabModeAsync
            $ \_ -> forever $ do
                Xlib.nextEvent display ev
                evType <- Xlib.get_EventType ev
                when (evType == Xlib.keyPress) $ killFirstNotification sTV


    -- New / Expired Checkers

    -- Push New Notifications from the DBus Queue
    void . GLib.timeoutAdd GLib.PRIORITY_DEFAULT 250 $ do
        mapM_ (showNotification sTV)
            . filter (\n -> nTitle n /= "" || nBody n /= "")
            <=< atomically $ do
                s <- readTVar sTV
                let ns = reverse $ appNotificationQueue s
                writeTVar sTV $ s { appNotificationQueue = [] }
                return ns
        return True

    -- Expire Any Timed Out Notifications
    void . GLib.timeoutAdd GLib.PRIORITY_DEFAULT 250 $ do
        currentTime <- getCurrentTime
        toDelete <- atomically $ do
            (expired, ok) <-
                partition (maybe False (< currentTime) . nExpirationTime . fst)
                . appWindowList <$> readTVar sTV
            modifyTVar sTV $ \s -> s { appWindowList = ok }
            return expired
        mapM_ (uncurry (deleteNotification sTV Expired) . first nID) toDelete
        return True

    -- Run the loop
    bracket (return shortcutThreadId) killThread $ const Gtk.main


-- | Grab a Specific Key Input, Run an Action & Then Ungrab the Input.
withGrabKey :: Xlib.Display
            -> Xlib.KeyCode
            -> Xlib.KeyMask
            -> Xlib.Window
            -> Bool
            -> Xlib.GrabMode
            -> Xlib.GrabMode
            -> (() -> IO ())
            -> IO ()
withGrabKey d k m w o p kb =
    bracket
        (Xlib.grabKey d k m w o p kb)
        (const $ Xlib.ungrabKey d k m w)

-- | Compose Multiple withGrabKey Calls Which Grab a Key for the Specified
-- KeyMask and Any Combination of CapsLock, NumLock, & ScrollLock.
--
-- E.g. Using this function to bind Ctrl-W to an action will bind Ctrl-W,
-- Caps-Ctrl-W, Num-Caps-Ctrl-W, Num-Scroll-Ctrl-W, etc. to the action.
grabKeyWithIgnoreMasks
    :: Xlib.Display
    -> Xlib.KeyCode
    -> Xlib.KeyMask
    -> Xlib.Window
    -> Bool
    -> Xlib.GrabMode
    -> Xlib.GrabMode
    -> (() -> IO ())
    -> IO ()
grabKeyWithIgnoreMasks d k m w o p kb action =
    let
        ignoredModifiers =
            [Xlib.lockMask, Xlib.mod2Mask, Xlib.mod3Mask]
        baseKeyFunction =
            const $ withGrabKey d k m w o p kb action
        withIgnoredKeyMasks = foldl ignoredGrab baseKeyFunction (subsequences ignoredModifiers)
    in
        withIgnoredKeyMasks ()
    where
        ignoredGrab :: (() -> IO ()) -> [Xlib.KeyMask] -> (() -> IO ())
        ignoredGrab f masks _ =
            withGrabKey d k (m .|. foldl (.|.) Xlib.noModMask masks) w o p kb f


getMonitorGeometryOrExit :: IO (Int32, Int32)
getMonitorGeometryOrExit = do
    maybeMonitorGeometry <- Gdk.displayOpen "" >>= \case
        Just d ->
            (Gdk.displayGetPrimaryMonitor d <|> Gdk.displayGetMonitor d 0)
            >>= (fmap join . traverse Gdk.getMonitorGeometry)
        Nothing ->
            putStrLn "Could not open display." >> exitFailure
    case maybeMonitorGeometry of
        Just g ->
             (,)
                <$> Gdk.getRectangleX g
                <*> Gdk.getRectangleY g
        Nothing ->
            putStrLn "Could not find monitor." >> exitFailure

-- | Create & Position a Notification Window & Attach the Click Handler.
--
-- Updates the `appNextPosition` & `appWindowList` state variables.
showNotification :: TVar AppState -> Notification -> IO ()
showNotification sTV n = do
    (winX, winY) <- appNextPosition <$> readTVarIO sTV
    win <- buildNotificationWindow n
    void . Gtk.onWidgetButtonPressEvent win . const
        $ deleteNotification sTV Dismissed (nID n) win
    Gtk.windowMove win winX winY
    Gtk.widgetShowAll win

    winHeight <- Gtk.widgetGetAllocatedHeight win
    atomically $ modifyTVar sTV $ \s -> s
        { appNextPosition =
            ( fst $ appNextPosition s
            , winHeight + snd (appNextPosition s) + notificationSpacing
            )
        , appWindowList =
            appWindowList s ++ [(n, win)]
        }


-- | TODO: Add Icons, Urgencies, Customizable Rendering
buildNotificationWindow :: Notification -> IO Gtk.Window
buildNotificationWindow n = do
    win <- Gtk.windowNew Gtk.WindowTypeToplevel
    Gtk.setWindowTypeHint win Gdk.WindowTypeHintNotification
    Gtk.windowSetKeepAbove win True
    Gtk.containerSetBorderWidth win 1

    -- Add Grid
    grid <- Gtk.gridNew
    Gtk.containerAdd win grid
    Gtk.orientableSetOrientation grid Gtk.OrientationVertical

    -- Set Notification Padding
    mapM_ (\f -> f grid verticalPadding)
        [ Gtk.widgetSetMarginTop
        , Gtk.widgetSetMarginBottom
        ]
    mapM_ (\f -> f grid horizontalPadding)
        [ Gtk.widgetSetMarginStart
        , Gtk.widgetSetMarginEnd
        ]

    -- Add Labels
    when (nTitle n /= "") $ do
        titleLabel <- Gtk.labelNew Nothing
        Gtk.labelSetMarkup titleLabel
            $ "<span color='#a6e22e' weight='bold'>" <> nTitle n <> "</span>"
        Gtk.widgetSetHalign titleLabel Gtk.AlignStart
        Gtk.containerAdd grid titleLabel

    when (nBody n /= "") $ do
        bodyLabel <- Gtk.labelNew Nothing
        Gtk.labelSetMarkup bodyLabel
            $ "<span color='#f8f8f0'>" <> nBody n <> "</span>"
        Gtk.widgetSetHalign bodyLabel Gtk.AlignStart
        Gtk.containerAdd grid bodyLabel

    return win

deleteNotification :: TVar AppState -> ReasonClosed -> NotificationID -> Gtk.Window -> IO Bool
deleteNotification sTV reason notificationID win =  do
    widgetHeight <- Gtk.widgetGetAllocatedHeight win
    atomically . modifyTVar sTV $ \s -> s
        { appWindowList =
            filter ((/= notificationID) . nID . fst) $ appWindowList s
        , appNextPosition =
            second (\y -> y - widgetHeight - notificationSpacing)
                $ appNextPosition s
        , appRemovalQueue =
            (notificationID, reason) : appRemovalQueue s
        }
    moveWindowsIfNecessary sTV
    Gtk.widgetDestroy win
    return True

killFirstNotification :: TVar AppState -> IO ()
killFirstNotification sTV = void . Gdk.threadsAddIdle GLib.PRIORITY_DEFAULT $
    listToMaybe . appWindowList <$> readTVarIO sTV >>= \case
        Just (n, win) ->
            deleteNotification sTV Dismissed (nID n) win >> return False
        Nothing ->
            return False


moveWindowsIfNecessary :: TVar AppState -> IO ()
moveWindowsIfNecessary sTV = do
    state <- readTVarIO sTV
    (_, windowsAndPositions) <- foldM nextPosition (appRootPosition state, [])
        $ appWindowList state
    mapM_ move windowsAndPositions
    where
        nextPosition ((x, y), processed) (_, win) = do
            widgetHeight <- Gtk.widgetGetAllocatedHeight win
            let nextY = y + widgetHeight + notificationSpacing
            return ((x, nextY), (win, (x, y)) : processed)
        move (win, (x, y)) = do
              currentY <- snd <$> Gtk.windowGetPosition win
              when (currentY /= y) $ Gtk.windowMove win x y




-- DBUS

-- | Connect to the DBus Session as the Notification Daemon & Handle the
-- Required Messages.
connectAndServe :: TVar AppState -> IO ()
connectAndServe sTV = bracket connectSession disconnect $ \client -> do
    requestResult <- requestName client "org.freedesktop.Notifications" []
    when (requestResult /= NamePrimaryOwner) $
        putStrLn "Another notification server is already started"
            >> exitFailure
    notificationServer sTV client
    forever $ do
        removals <- atomically $ do
            nis <- appRemovalQueue <$> readTVar sTV
            modifyTVar sTV $ \s -> s { appRemovalQueue = [] }
            return nis
        mapM_ (uncurry $ emitClosed client) removals
        threadDelay 1000000

-- | Implement the `org.freedesktop.Notifications` DBus interface at a path
-- of `/org/freedesktop/Notifications`.
notificationServer :: TVar AppState -> Client -> IO ()
notificationServer sTV client =
    export client "/org/freedesktop/Notifications"
        [ autoMethod "org.freedesktop.Notifications" "GetCapabilities" getCapabilities
        , autoMethod "org.freedesktop.Notifications" "Notify" (notify sTV)
        , autoMethod "org.freedesktop.Notifications" "CloseNotification" closeNotification
        , autoMethod "org.freedesktop.Notifications" "GetServerInformation" getServerInformation
        ]


-- DBus Messages

getCapabilities :: IO [String]
getCapabilities = return
    [ "body"
    , "body-hyperlinks"
    , "body-markup"
    , "persistence"
    ]


-- implement Notify
-- TODO: Implement replace
notify
    :: TVar AppState            -- ^ App State
    -> String                   -- ^ App Name
    -> Word32                   -- ^ ID to Replace (if > 0)
    -> String                   -- ^ App Icon
    -> T.Text                   -- ^ Summary
    -> T.Text                   -- ^ Body
    -> [String]                 -- ^ Actions
    -> M.Map String Variant     -- ^ Hint
    -> Int32                    -- ^ Timeout in Milliseconds
    -> IO Word32                -- ^ Notification ID
notify sTV _ _ _ summary body _ _ timeout = do
    expirationTime <-
        if timeout > 0 then
            Just . addUTCTime (fromIntegral timeout / 1000) <$> getCurrentTime
        else
            return Nothing
    atomically $ do
        state <- readTVar sTV
        let notificationID =
                appNextNotificationID state
            notification =
                Notification
                    { nID = notificationID
                    , nBody = body
                    , nTitle = summary
                    , nExpirationTime = expirationTime
                    }
            updatedState =
                state
                    { appNotificationQueue = notification : appNotificationQueue state
                    , appNextNotificationID = nextID notificationID
                    }
        writeTVar sTV updatedState
        return $ fromNotificationID notificationID

-- TODO: implement CloseNotification
closeNotification :: Word32 -> IO ()
closeNotification w =
    print $ "Closes Notification #" ++ show w

-- Implement GetServerInformation
getServerInformation :: IO (String, String, String, String)
getServerInformation = return
    ( "hs-notifications"
    , "prikhi"
    , "0.1.0.0"
    , "1.2"
    )


-- Signals(send)

-- implement NotificationClosed
data ReasonClosed
    = Expired
    | Dismissed
    | DBusCall
    | Other

instance IsVariant ReasonClosed where
    toVariant a = toVariant $ case a of
        Expired -> 1 :: Word32
        Dismissed -> 2
        DBusCall -> 3
        Other -> 4
    fromVariant = (fromVariant :: Variant -> Maybe Word32) >=> \case
        1 -> Just Expired
        2 -> Just Dismissed
        3 -> Just DBusCall
        4 -> Just Other
        _ -> Nothing

emitClosed :: Client -> NotificationID -> ReasonClosed -> IO ()
emitClosed c (NotificationID i) r =
    emit c $ (signal "/org/freedesktop/Notifications" "org.freedesktop.Notifications" "NotificationClosed")
        { signalBody = [toVariant i, toVariant r]
        }


-- Optionally Implement ActionInvoked


src/HsNotifications.hs
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE OverloadedStrings #-}
module HsNotifications where

import Control.Applicative ((<|>))
import Control.Arrow (first, second)
import Control.Concurrent (threadDelay, forkIO, killThread)
import Control.Concurrent.STM (TVar, atomically, newTVarIO, modifyTVar, readTVar, readTVarIO, writeTVar)
import Control.Exception (bracket)
import Control.Monad ((<=<), forever, when, void, join, foldM)
import Data.Int (Int32)
import Data.List (partition, find)
import Data.Maybe (listToMaybe, fromMaybe)
import Data.Text.Encoding (encodeUtf8)
import Data.Time.Clock (getCurrentTime, addUTCTime)
import Data.Word (Word32)
import Stitch ((?), (.=), renderCSS)
import System.Exit (exitFailure)

import DBus
import DBus.Client

import HsNotifications.Config (Config(..))
import HsNotifications.Models
import HsNotifications.Shortcuts (withShortcutThread)

import qualified Data.Map as M
import qualified Data.Text as T
import qualified GI.Gtk as Gtk
import qualified GI.Gdk as Gdk
import qualified GI.GLib as GLib


-- | Initialize the `AppState`, fork the DBus client & run the GTK app.
--
-- TODO: `Config` should be a `ReaderT`, `TVar AppState` should be
-- a `StateT`, transformer stack should implement `MonadIO`.
run :: Config -> IO ()
run config =
    initialState config >>= \s -> forkIO (connectAndServe config s) >> runGtk config s

-- | Create a new `TVar` holding the initial application state.
--
-- TODO: It'd probably be better to limit what is available between GTK
-- & DBus instead of passing the whole state around.
initialState :: Config -> IO (TVar AppState)
initialState c = do
    rootPosition <- first (+ placementX c) . second (+ placementY c)
        <$> getMonitorGeometryOrExit
    newTVarIO AppState
        { appNotificationQueue =
            []
        , appRemovalQueue =
            []
        , appWindowList =
            []
        , appRootPosition =
            rootPosition
        , appNextPosition =
            rootPosition
        , appNextNotificationID =
            NotificationID minBound
        }

-- | Get the X & Y Coordinates of the Primary `Monitor` on the default
-- `Display`.
--
-- If the Primary Monitor cannot be identified, it uses the first monitor
-- of the `Display`.
--
-- Exits with `exitFailure` if the Display could not be opened or a Monitor
-- could not be found.
getMonitorGeometryOrExit :: IO (Int32, Int32)
getMonitorGeometryOrExit = do
    maybeMonitorGeometry <- Gdk.displayOpen "" >>= \case
        Just d ->
            (Gdk.displayGetPrimaryMonitor d <|> Gdk.displayGetMonitor d 0)
            >>= (fmap join . traverse Gdk.getMonitorGeometry)
        Nothing ->
            putStrLn "Could not open display." >> exitFailure
    case maybeMonitorGeometry of
        Just g ->
             (,)
                <$> Gdk.getRectangleX g
                <*> Gdk.getRectangleY g
        Nothing ->
            putStrLn "Could not find monitor." >> exitFailure


-- TODO: Make left & right notification lists
runGtk :: Config -> TVar AppState -> IO ()
runGtk c sTV = do
    initializeGtk c

    -- Keybind Watchers
    closeOneShortcutThread <-
        withShortcutThread (closeKey c) (closeSingleMask c)
            $ killFirstNotification c sTV
    closeAllShortcutThread <-
        withShortcutThread (closeKey c) (closeAllMask c)
            $ killAllNotifications c sTV

    -- New / Expired Checkers
    void . GLib.timeoutAdd GLib.PRIORITY_DEFAULT 100
        $ processNotificationQueue c sTV
            >> removeExpired c sTV
            >> return True

    -- Run the loop
    bracket (return [closeOneShortcutThread, closeAllShortcutThread]) (mapM killThread)
        $ const Gtk.main


-- | Initialize GTK & Attach the Generated CSS to the Default `Gdk.Screen`.
initializeGtk :: Config -> IO ()
initializeGtk c =
    Gtk.init Nothing >> Gdk.screenGetDefault >>= \case
        Nothing ->
            return ()
        Just s -> do
            provider <- Gtk.cssProviderNew
            Gtk.cssProviderLoadFromData provider . encodeUtf8 $ appStyle c
            Gtk.styleContextAddProviderForScreen s provider 600


-- | Generate the CSS used to theme the Windows.
appStyle :: Config -> T.Text
appStyle c = renderCSS $
    "window" ? do
        when (font c /= "") $
            "font" .= font c
        "background-color" .= backgroundColor c
        "border-style" .= "solid"
        "border-width" .= "1px"
        "border-color" .= borderColor c


-- | Either Handle a QueueRequest for a Notification by Showing or
-- Replacing It.
handleQueueRequest :: Config -> TVar AppState -> Notification -> QueueRequest -> IO ()
handleQueueRequest c sTV n = \case
    Add ->
        when (nTitle n /= "" || nBody n /= "") $ showNotification c sTV n
    Replace ->
        replaceNotification c sTV n


-- | Push New Notifications from the DBus Queue.
--
-- Thread-safe.
processNotificationQueue :: Config -> TVar AppState -> IO ()
processNotificationQueue c sTV = void . Gdk.threadsAddIdle GLib.PRIORITY_DEFAULT $ do
    mapM_ (uncurry $ handleQueueRequest c sTV) <=< atomically $ do
        s <- readTVar sTV
        let ns = reverse $ appNotificationQueue s
        writeTVar sTV $ s { appNotificationQueue = [] }
        return ns
    return False


-- | Expire Any Timed Out Notifications
--
-- Thread-safe.
removeExpired :: Config -> TVar AppState -> IO ()
removeExpired c sTV = void . Gdk.threadsAddIdle GLib.PRIORITY_DEFAULT $ do
    currentTime <- getCurrentTime
    toDelete <- atomically $ do
        (expired, ok) <-
            partition (maybe False (< currentTime) . nExpirationTime . fst)
            . appWindowList <$> readTVar sTV
        modifyTVar sTV $ \s -> s { appWindowList = ok }
        return expired
    mapM_ (uncurry (deleteNotification c sTV Expired) . first nID) toDelete
    return False


-- | Create & Position a Notification Window & Attach the Click Handler.
--
-- Updates the `appNextPosition` & `appWindowList` state variables.
showNotification :: Config -> TVar AppState -> Notification -> IO ()
showNotification c sTV n = do
    (winX, winY) <- appNextPosition <$> readTVarIO sTV
    win <- buildNotificationWindow c n
    void . Gtk.onWidgetButtonPressEvent win . const
        $ deleteNotification c sTV Dismissed (nID n) win
    Gtk.windowMove win winX winY
    Gtk.widgetShowAll win

    winHeight <- Gtk.widgetGetAllocatedHeight win
    atomically $ modifyTVar sTV $ \s -> s
        { appNextPosition =
            ( fst $ appNextPosition s
            , winHeight + snd (appNextPosition s) + spacing c
            )
        , appWindowList =
            appWindowList s ++ [(n, win)]
        }


-- | Create & fill a new `Gtk.Window` for the `Notification`.
--
-- The window contains a `Gtk.Grid` with `Gtk.Label`s for the title & body
-- text.
--
-- TODO: Add Icons
buildNotificationWindow :: Config -> Notification -> IO Gtk.Window
buildNotificationWindow c n = do
    win <- Gtk.windowNew Gtk.WindowTypeToplevel
    Gtk.setWindowTypeHint win Gdk.WindowTypeHintNotification
    Gtk.windowSetKeepAbove win True
    Gtk.containerSetBorderWidth win 1

    -- Add Grid
    grid <- Gtk.gridNew
    Gtk.containerAdd win grid
    Gtk.orientableSetOrientation grid Gtk.OrientationVertical

    -- Set Padding
    mapM_ (\f -> f grid $ paddingY c)
        [ Gtk.widgetSetMarginTop
        , Gtk.widgetSetMarginBottom
        ]
    mapM_ (\f -> f grid $ paddingX c)
        [ Gtk.widgetSetMarginStart
        , Gtk.widgetSetMarginEnd
        ]

    -- Add Labels
    when (nTitle n /= "") $ do
        titleLabel <- Gtk.labelNew Nothing
        Gtk.labelSetMarkup titleLabel $ titleFormat c n
        Gtk.widgetSetHalign titleLabel Gtk.AlignStart
        Gtk.containerAdd grid titleLabel

    when (nBody n /= "") $ do
        bodyLabel <- Gtk.labelNew Nothing
        Gtk.labelSetMarkup bodyLabel . bodyFormat c $ nBody n
        Gtk.widgetSetHalign bodyLabel Gtk.AlignStart
        Gtk.containerAdd grid bodyLabel

    return win


-- | Replace & Re-Render a Notification.
--
-- This replaces the Notification in the `WindowList` & updates the
-- `Gtk.Window`'s `Gtk.Label`s.
--
-- TODO: If a body label is missing but the new notification has body text,
-- we should create and add one.
--
--  Thread-safe.
replaceNotification :: Config -> TVar AppState -> Notification -> IO ()
replaceNotification c sTV newN = void . Gdk.threadsAddIdle GLib.PRIORITY_DEFAULT $ do
    maybeWindow <- atomically $ do
        windowList <- appWindowList <$> readTVar sTV
        let (newList, maybeWindow) = replace newN ([], Nothing) windowList
        modifyTVar sTV $ \s -> s { appWindowList = newList }
        return maybeWindow
    maybeM maybeWindow $ \win -> do
        maybeGrid <- join . listToMaybe <$> (mapM (Gtk.castTo Gtk.Container) =<< Gtk.containerGetChildren win)
        maybeM maybeGrid $ \grid -> do
            labels <- mapM (Gtk.castTo Gtk.Label) =<< Gtk.containerGetChildren grid
            case labels of
                (title : rest) -> do
                    maybeM title $ \l ->
                        Gtk.labelSetMarkup l $ titleFormat c newN
                    maybeM (join $ listToMaybe rest) $ \l ->
                        Gtk.labelSetMarkup l . bodyFormat c $ nBody newN
                _ ->
                    return ()
    moveWindowsIfNecessary c sTV
    return False
    where maybeM ma f =
            maybe (return ()) f ma
          replace v (processed, mWin) xs =
            case xs of
                [] ->
                    (processed, mWin)
                (n, w):rest ->
                    if nID n == nID v then
                        (processed ++ [(v, w)] ++ rest, Just w)
                    else
                        replace v (processed ++ [(n, w)], mWin) rest


-- | Remove a `Notification`.
--
-- Destroy's the `Gtk.Window`, removes it from the `WindowList`, & adds it
-- to the `appRemovalQueue`.
deleteNotification :: Config -> TVar AppState -> ReasonClosed -> NotificationID -> Gtk.Window -> IO Bool
deleteNotification c sTV reason notificationID win =  do
    widgetHeight <- Gtk.widgetGetAllocatedHeight win
    atomically . modifyTVar sTV $ \s -> s
        { appWindowList =
            filter ((/= notificationID) . nID . fst) $ appWindowList s
        , appNextPosition =
            second (\y -> y - widgetHeight - spacing c)
                $ appNextPosition s
        , appRemovalQueue =
            (notificationID, reason) : appRemovalQueue s
        }
    moveWindowsIfNecessary c sTV
    Gtk.widgetDestroy win
    return True


-- | Remove the first `Notification` in the `WindowList`.
--
-- Thread-safe.
killFirstNotification :: Config -> TVar AppState -> IO ()
killFirstNotification c sTV = void . Gdk.threadsAddIdle GLib.PRIORITY_DEFAULT $
    listToMaybe . appWindowList <$> readTVarIO sTV >>= \case
        Just (n, win) ->
            deleteNotification c sTV Dismissed (nID n) win >> return False
        Nothing ->
            return False

-- | Remove all `Notification`s currently in the `WindowList`.
--
--  Thread-safe.
killAllNotifications :: Config -> TVar AppState -> IO ()
killAllNotifications c sTV = void . Gdk.threadsAddIdle GLib.PRIORITY_DEFAULT $
    appWindowList <$> readTVarIO sTV
        >>= mapM_ (uncurry (deleteNotification c sTV Dismissed) . first nID)
        >> return False

-- | Remove the `Notification` with the given `NotificationID`.
--
-- Thread-safe.
killNotificationByID :: Config -> TVar AppState -> ReasonClosed -> NotificationID -> IO ()
killNotificationByID c sTV reason notifID = void . Gdk.threadsAddIdle GLib.PRIORITY_DEFAULT $
    maybe (return False) (\(_, w) -> deleteNotification c sTV reason notifID w >> return False)
    =<< find ((== notifID) . nID . fst) . appWindowList <$> readTVarIO sTV


-- | Calculate the Actual & Expected Position of the Notifications & Move
-- Them If Necessary.
moveWindowsIfNecessary :: Config -> TVar AppState -> IO ()
moveWindowsIfNecessary c sTV = do
    state <- readTVarIO sTV
    (_, windowsAndPositions) <- foldM nextPosition (appRootPosition state, [])
        $ appWindowList state
    mapM_ move windowsAndPositions
    where
        nextPosition ((x, y), processed) (_, win) = do
            widgetHeight <- Gtk.widgetGetAllocatedHeight win
            let nextY = y + widgetHeight + spacing c
            return ((x, nextY), (win, (x, y)) : processed)
        move (win, (x, y)) = do
              currentY <- snd <$> Gtk.windowGetPosition win
              when (currentY /= y) $ Gtk.windowMove win x y




-- DBUS

-- | Connect to the DBus Session as the Notification Daemon & Handle the
-- Required Messages.
connectAndServe :: Config -> TVar AppState -> IO ()
connectAndServe c sTV = bracket connectSession disconnect $ \client -> do
    requestResult <- requestName client "org.freedesktop.Notifications" []
    when (requestResult /= NamePrimaryOwner) $
        putStrLn "Another notification server is already started"
            >> exitFailure
    notificationServer c sTV client
    forever $ do
        removals <- atomically $ do
            nis <- appRemovalQueue <$> readTVar sTV
            modifyTVar sTV $ \s -> s { appRemovalQueue = [] }
            return nis
        mapM_ (uncurry $ emitClosed client) removals
        threadDelay 1000000

-- | Implement the `org.freedesktop.Notifications` DBus interface at a path
-- of `/org/freedesktop/Notifications`.
notificationServer :: Config -> TVar AppState -> Client -> IO ()
notificationServer c sTV client =
    export client "/org/freedesktop/Notifications"
        [ autoMethod "org.freedesktop.Notifications" "GetCapabilities" getCapabilities
        , autoMethod "org.freedesktop.Notifications" "Notify" (notify sTV)
        , autoMethod "org.freedesktop.Notifications" "CloseNotification" (closeNotification c sTV)
        , autoMethod "org.freedesktop.Notifications" "GetServerInformation" getServerInformation
        ]


-- DBus Messages

-- | Return the server's capabilities.
getCapabilities :: IO [String]
getCapabilities = return
    [ "body"
    , "body-hyperlinks"
    , "body-markup"
    , "persistence"
    ]


-- | Determine whether the Notification is New or a Replacement & add the
-- request to the `NotificationQueue`.
notify
    :: TVar AppState            -- ^ App State
    -> String                   -- ^ App Name
    -> Word32                   -- ^ ID to Replace (if > 0)
    -> String                   -- ^ App Icon
    -> T.Text                   -- ^ Summary
    -> T.Text                   -- ^ Body
    -> [String]                 -- ^ Actions
    -> M.Map String Variant     -- ^ Hint
    -> Int32                    -- ^ Timeout in Milliseconds
    -> IO Word32                -- ^ Notification ID
notify sTV _ replaceID _ summary body _ hints timeout = do
    expirationTime <-
        if timeout > 0 then
            Just . addUTCTime (fromIntegral timeout / 1000) <$> getCurrentTime
        else
            return Nothing
    atomically $ do
        state <- readTVar sTV
        let (notificationID, nextNotificationID, queueRequest) =
                if replaceID == 0 then
                    ( appNextNotificationID state
                    , nextID $ appNextNotificationID state
                    , Add
                    )
                else
                    ( NotificationID replaceID
                    , appNextNotificationID state
                    , Replace
                    )
            urgency =
                fromMaybe Normal $ fromVariant =<< M.lookup "urgency" hints
            notification =
                Notification
                    { nID = notificationID
                    , nBody = body
                    , nTitle = summary
                    , nUrgency = urgency
                    , nExpirationTime = expirationTime
                    }
            updatedState =
                state
                    { appNotificationQueue = (notification, queueRequest) : appNotificationQueue state
                    , appNextNotificationID = nextNotificationID
                    }
        writeTVar sTV updatedState
        return $ fromNotificationID notificationID


-- | The CloseNotification DBus method removes the `Window` representing
-- the `Notification`.
closeNotification :: Config -> TVar AppState -> Word32 -> IO ()
closeNotification c sTV =
    killNotificationByID c sTV DBusCall . NotificationID


-- | Return the Sever's Name, Vendor, Version, & Spec Version.
getServerInformation :: IO (String, String, String, String)
getServerInformation = return
    ( "hs-notifications"
    , "prikhi"
    , "0.1.0.0"
    , "1.2"
    )


-- Signal Emission
-- TODO: Optionally Implement ActionInvoked

-- | Notify DBus Listeners that a Notification has been Closed.
emitClosed :: Client -> NotificationID -> ReasonClosed -> IO ()
emitClosed c (NotificationID i) r =
    emit c $ (signal "/org/freedesktop/Notifications" "org.freedesktop.Notifications" "NotificationClosed")
        { signalBody = [toVariant i, toVariant r]
        }
